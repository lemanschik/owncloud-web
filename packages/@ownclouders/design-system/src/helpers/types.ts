export type IconFillType = 'fill' | 'line' | 'none'
