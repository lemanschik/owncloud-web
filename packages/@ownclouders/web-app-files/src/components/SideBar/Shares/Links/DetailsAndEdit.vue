<template>
  <div class="link-details oc-flex oc-flex-between oc-flex-middle">
    <div v-if="isModifiable">
      <oc-button
        :id="`edit-public-link-role-dropdown-toggle-${link.id}`"
        appearance="raw"
        class="edit-public-link-role-dropdown-toggle oc-text-left"
        gap-size="none"
      >
        <span class="link-current-role" v-text="currentLinkRoleLabel" />
        <oc-icon name="arrow-down-s" />
      </oc-button>
      <oc-drop
        ref="editPublicLinkRoleDropdown"
        class="edit-public-link-role-dropdown"
        :drop-id="`edit-public-link-role-dropdown`"
        :toggle="`#edit-public-link-role-dropdown-toggle-${link.id}`"
        padding-size="small"
        offset="0"
        mode="click"
      >
        <oc-list class="role-dropdown-list">
          <li v-for="roleOption in availableRoleOptions" :key="`role-dropdown-${roleOption.key}`">
            <oc-button
              :id="`files-role-${roleOption.name}`"
              :class="{
                selected: isSelectedRole(roleOption),
                'oc-background-primary-gradient': isSelectedRole(roleOption)
              }"
              :appearance="isSelectedRole(roleOption) ? 'raw-inverse' : 'raw'"
              :variation="isSelectedRole(roleOption) ? 'primary' : 'passive'"
              justify-content="space-between"
              class="oc-p-s"
              @click="
                updateLink({
                  link: {
                    ...link,
                    permissions: roleOption.bitmask(false)
                  },
                  dropRef: $refs.editPublicLinkRoleDropdown
                })
              "
            >
              <span class="oc-flex oc-flex-middle">
                <oc-icon :name="roleOption.icon" class="oc-pl-s oc-pr-m" variation="inherit" />
                <span>
                  <span
                    class="oc-text-bold oc-display-block oc-width-1-1"
                    v-text="$gettext(roleOption.label)"
                  />
                  <span class="oc-text-small">{{ $gettext(roleOption.description(false)) }}</span>
                </span>
              </span>
              <span class="oc-flex">
                <oc-icon v-if="isSelectedRole(roleOption)" name="check" variation="inherit" />
              </span>
            </oc-button>
          </li>
        </oc-list>
      </oc-drop>
    </div>
    <p v-else class="oc-my-rm">
      <span
        v-oc-tooltip="$gettext(currentLinkRoleDescription)"
        class="link-current-role"
        v-text="$gettext(currentLinkRoleLabel)"
      />
    </p>
    <div :class="{ 'oc-pr-s': !isModifiable }" class="details-buttons">
      <oc-button
        v-if="link.indirect"
        v-oc-tooltip="viaTooltip"
        :area-label="viaTooltip"
        type="router-link"
        class="oc-files-file-link-via"
        appearance="raw"
        :to="viaRouterParams"
      >
        <oc-icon name="folder-shared" fill-type="line" />
      </oc-button>
      <oc-icon
        v-if="link.password"
        v-oc-tooltip="passwortProtectionTooltip"
        name="lock-password"
        fill-type="line"
        :aria-label="passwortProtectionTooltip"
      />
      <oc-icon
        v-if="link.expiration"
        v-oc-tooltip="expirationDateTooltip"
        class="oc-files-public-link-expires oc-ml-xs"
        :data-testid="`files-link-id-${link.id}-expiration-date`"
        :aria-label="expirationDateTooltip"
        name="calendar-event"
        fill-type="line"
      />
      <div v-if="isModifiable">
        <oc-button
          :id="`edit-public-link-dropdown-toggl-${link.id}`"
          appearance="raw"
          class="edit-drop-trigger"
          :data-testid="`files-link-id-${link.id}-btn-edit`"
        >
          <oc-icon name="more-2" />
        </oc-button>
        <oc-drop
          ref="editPublicLinkDropdown"
          :drop-id="`edit-public-link-dropdown`"
          :toggle="`#edit-public-link-dropdown-toggl-${link.id}`"
          padding-size="small"
          mode="click"
        >
          <oc-list class="edit-public-link-dropdown-menu oc-files-context-actions-border oc-pb-s">
            <li
              v-for="(option, i) in editOptions"
              :key="`public-link-edit-option-${i}`"
              class="oc-rounded oc-menu-item-hover"
            >
              <oc-datepicker
                v-if="option.showDatepicker"
                v-model="newExpiration"
                class="link-expiry-picker"
                :min-date="expirationDate.min"
                :max-date="expirationDate.max"
                :locale="$language.current"
                :is-required="expirationDate.enforce"
              >
                <template #default="{ togglePopover }">
                  <oc-button
                    :data-testid="`files-link-id-${link.id}-edit-${option.id}`"
                    appearance="raw"
                    class="oc-p-s action-menu-item"
                    :variation="option.variation"
                    @click="togglePopover"
                  >
                    <oc-icon :name="option.icon" fill-type="line" size="medium" />
                    <span v-text="option.title" />
                  </oc-button>
                </template>
              </oc-datepicker>
              <oc-button
                v-else
                appearance="raw"
                class="oc-p-s action-menu-item"
                :data-testid="`files-link-id-${link.id}-edit-${option.id}`"
                @click="option.method"
              >
                <oc-icon :name="option.icon" fill-type="line" size="medium" />
                <span v-text="option.title" />
              </oc-button>
            </li>
          </oc-list>
          <oc-list class="edit-public-link-dropdown-menu oc-pt-s">
            <li class="oc-rounded oc-menu-item-hover">
              <oc-button
                appearance="raw"
                class="oc-p-s action-menu-item"
                :data-testid="`files-link-id-${link.id}-edit-${deleteOption.id}`"
                @click="deleteOption.method"
              >
                <oc-icon :name="deleteOption.icon" fill-type="line" size="medium" />
                <span v-text="deleteOption.title" />
              </oc-button>
            </li>
          </oc-list>
        </oc-drop>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { basename } from 'path'
import { DateTime } from 'luxon'
import { mapActions, mapGetters } from 'vuex'
import { createLocationSpaces } from '../../../../router'
import {
  linkRoleInternalFile,
  linkRoleInternalFolder,
  LinkShareRoles,
  ShareRole
} from 'web-client/src/helpers/share'
import { defineComponent, inject, PropType, Ref } from 'vue'
import { formatDateFromDateTime, formatRelativeDateFromDateTime } from 'web-pkg/src/helpers'
import { Resource, SpaceResource } from 'web-client/src/helpers'
import { createFileRouteOptions } from 'web-pkg/src/helpers/router'
import { OcDrop } from 'design-system/src/components'

export default defineComponent({
  name: 'DetailsAndEdit',
  props: {
    availableRoleOptions: {
      type: Array as PropType<ShareRole[]>,
      required: true
    },
    canRename: {
      type: Boolean,
      default: false
    },
    expirationDate: {
      type: Object,
      default: () => ({}),
      required: true
    },
    isFolderShare: {
      type: Boolean,
      default: false
    },
    isModifiable: {
      type: Boolean,
      default: false
    },
    isPasswordEnforced: {
      type: Boolean,
      default: false
    },
    link: {
      type: Object,
      required: true
    }
  },
  emits: ['removePublicLink', 'updateLink'],
  setup() {
    return {
      space: inject<Ref<SpaceResource>>('space'),
      resource: inject<Ref<Resource>>('resource')
    }
  },
  data() {
    return {
      newExpiration: this.link.expiration
    }
  },
  computed: {
    ...mapGetters('runtime/spaces', ['spaces']),
    currentLinkRole() {
      return LinkShareRoles.getByBitmask(this.link.permissions, this.isFolderShare)
    },
    currentLinkRoleDescription() {
      return this.currentLinkRole.description(false)
    },

    currentLinkRoleLabel() {
      if (this.currentLinkRole.longLabel !== '') {
        return this.$gettext(this.currentLinkRole.longLabel)
      } else {
        return this.$gettext(this.currentLinkRole.label)
      }
    },

    editOptions() {
      const result = []

      if (this.canRename) {
        result.push({
          id: 'rename',
          title: this.$gettext('Rename'),
          icon: 'pencil',
          method: this.showRenameModal
        })
      }

      if (this.link.expiration) {
        result.push({
          id: 'edit-expiration',
          title: this.$gettext('Edit expiration date'),
          method: this.updateLink,
          icon: 'calendar-event',
          showDatepicker: true
        })
        if (!this.expirationDate.enforced) {
          result.push({
            id: 'remove-expiration',
            title: this.$gettext('Remove expiration date'),
            icon: 'calendar',
            method: () =>
              this.updateLink({
                link: {
                  ...this.link,
                  expiration: ''
                }
              })
          })
        }
      } else {
        result.push({
          id: 'add-expiration',
          title: this.$gettext('Add expiration date'),
          method: this.updateLink,
          icon: 'calendar-event',
          showDatepicker: true
        })
      }

      if (this.link.password) {
        result.push({
          id: 'edit-password',
          title: this.$gettext('Edit password'),
          icon: 'lock-password',
          method: this.showPasswordModal
        })

        if (!this.isPasswordEnforced) {
          result.push({
            id: 'remove-password',
            title: this.$gettext('Remove password'),
            icon: 'lock-unlock',
            method: () =>
              this.updateLink({
                link: {
                  ...this.link,
                  password: ''
                }
              })
          })
        }
      }
      if (!this.isPasswordEnforced && !this.link.password && !this.isAliasLink) {
        result.push({
          id: 'add-password',
          title: this.$gettext('Add password'),
          icon: 'lock-password',
          method: this.showPasswordModal
        })
      }

      return result
    },

    deleteOption() {
      return {
        id: 'delete',
        title: this.$gettext('Delete link'),
        method: this.deleteLink,
        icon: 'delete-bin-5',
        variation: 'danger'
      }
    },

    viaRouterParams() {
      const matchingSpace = (this.space ||
        this.spaces.find((space) => space.id === this.resource.storageId)) as SpaceResource
      if (!matchingSpace) {
        return {}
      }

      return createLocationSpaces(
        'files-spaces-generic',
        createFileRouteOptions(matchingSpace, {
          path: this.link.path,
          fileId: this.link.file.source
        })
      )
    },

    localExpirationDate() {
      return formatDateFromDateTime(
        DateTime.fromISO(this.link.expiration).endOf('day'),
        this.$language.current
      )
    },

    expirationDateRelative() {
      return formatRelativeDateFromDateTime(
        DateTime.fromISO(this.link.expiration).endOf('day'),
        this.$language.current
      )
    },

    expirationDateTooltip() {
      return this.$gettext(
        'Expires %{timeToExpiry} (%{expiryDate})',
        { timeToExpiry: this.expirationDateRelative, expiryDate: this.localExpirationDate },
        true
      )
    },

    viaTooltip() {
      if (!this.link.indirect) {
        return null
      }
      return this.$gettextInterpolate(
        this.$gettext('Navigate to the parent (%{folderName})'),
        { folderName: basename(this.link.path) },
        true
      )
    },

    passwortProtectionTooltip() {
      return this.$gettext('This link is password-protected')
    },

    isAliasLink() {
      return [linkRoleInternalFolder, linkRoleInternalFile].includes(this.currentLinkRole)
    }
  },
  watch: {
    newExpiration(expiration) {
      this.updateLink({
        link: {
          ...this.link,
          expiration
        }
      })
    }
  },
  methods: {
    ...mapActions([
      'createModal',
      'hideModal',
      'setModalInputErrorMessage',
      'setModalConfirmButtonDisabled'
    ]),

    isSelectedRole(role: ShareRole) {
      return this.link.permissions === role.bitmask(false)
    },
    updateLink({ link, dropRef = undefined, onSuccess = () => {} }) {
      link = link || this.link
      dropRef = dropRef || this.$refs.editPublicLinkDropdown
      this.$emit('updateLink', { link, onSuccess })
      dropRef.hide()
    },
    deleteLink() {
      this.$emit('removePublicLink', { link: this.link })
      ;(this.$refs.editPublicLinkDropdown as InstanceType<typeof OcDrop>).hide()
    },
    checkInputValue(value) {
      if (value.length > 255) {
        return this.setModalInputErrorMessage(
          this.$gettext('Link name cannot exceed 255 characters')
        )
      }
      this.setModalInputErrorMessage(null)
    },
    showRenameModal() {
      const modal = {
        variation: 'passive',
        title: this.$gettext('Edit name'),
        cancelText: this.$gettext('Cancel'),
        confirmText: this.$gettext('Save'),
        hasInput: true,
        inputValue: this.link.name,
        inputLabel: this.$gettext('Link name'),
        onCancel: this.hideModal,
        onInput: this.checkInputValue,
        onConfirm: (name) =>
          this.updateLink({
            link: {
              ...this.link,
              name
            },
            onSuccess: () => {
              this.hideModal()
            }
          })
      }

      this.createModal(modal)
    },

    checkPassword(password) {
      if (password === '') {
        this.setModalConfirmButtonDisabled(true)
        return this.setModalInputErrorMessage(this.$gettext("Password can't be empty"))
      }
      if (password.length > 72) {
        this.setModalConfirmButtonDisabled(true)
        return this.setModalInputErrorMessage(this.$gettext("Password can't exceed 72 characters"))
      }
      this.setModalConfirmButtonDisabled(false)
      return this.setModalInputErrorMessage(null)
    },

    showPasswordModal() {
      const modal = {
        variation: 'passive',
        title: this.link.password ? this.$gettext('Edit password') : this.$gettext('Add password'),
        cancelText: this.$gettext('Cancel'),
        confirmText: this.link.password ? this.$gettext('Apply') : this.$gettext('Set'),
        hasInput: true,
        confirmDisabled: true,
        inputLabel: this.$gettext('Password'),
        inputType: 'password',
        onCancel: this.hideModal,
        onInput: (password) => this.checkPassword(password),
        onConfirm: (password) => {
          this.updateLink({
            link: {
              ...this.link,
              password
            },
            onSuccess: () => {
              this.hideModal()
            }
          })
        }
      }
      this.createModal(modal)
    }
  }
})
</script>

<style lang="scss" scoped>
.details-buttons {
  min-width: 5rem !important;
  display: flex;
  justify-content: flex-end;
}

.edit-public-link-role-dropdown {
  width: 400px;
}

.role-dropdown-list span {
  line-height: 1.3;
}

.role-dropdown-list li {
  margin: var(--oc-space-xsmall) 0;

  &:first-child {
    margin-top: 0;
  }
  &:last-child {
    margin-bottom: 0;
  }

  .oc-button {
    text-align: left;
    width: 100%;
    gap: var(--oc-space-medium);

    &:hover,
    &:focus {
      background-color: var(--oc-color-background-hover);
      text-decoration: none;
    }
  }
}

.edit-public-link-dropdown-menu {
  .action-menu-item {
    width: 100%;
    justify-content: flex-start;
  }
}

.oc-files-public-link-expires {
  margin-top: 1px; // to align with the other elements
}
</style>
