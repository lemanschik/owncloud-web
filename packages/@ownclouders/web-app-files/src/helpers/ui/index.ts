export * as fileList from './filesList'
export * as resourceTable from './resourceTable'
export * as resourceTiles from './resourceTiles'
