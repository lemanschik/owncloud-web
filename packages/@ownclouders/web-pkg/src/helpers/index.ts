export * from './cache'
export * from './contextMenuDropdown'
export * from './download'
export * from './datetime'
export * from './filesize'
export * from './fuse'
export * from './locale'
