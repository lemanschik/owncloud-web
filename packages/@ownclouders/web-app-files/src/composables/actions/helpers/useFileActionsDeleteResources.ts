import { Store } from 'vuex'
import { cloneStateObject } from '../../../helpers/store'
import { isSameResource } from '../../../helpers/resource'
import { buildWebDavFilesTrashPath } from '../../../helpers/resources'
import { buildWebDavSpacesTrashPath, Resource, SpaceResource } from 'web-client/src/helpers'
import PQueue from 'p-queue'
import { isLocationTrashActive, isLocationSpacesActive } from '../../../router'
import { dirname } from 'path'
import { createFileRouteOptions } from 'web-pkg/src/helpers/router'
import { computed, unref } from 'vue'
import {
  useCapabilitySpacesEnabled,
  useCapabilityShareJailEnabled,
  useClientService,
  useRouter,
  useStore,
  useLoadingService
} from 'web-pkg/src/composables'
import { useGettext } from 'vue3-gettext'
import { ref } from 'vue'

export const useFileActionsDeleteResources = ({ store }: { store?: Store<any> }) => {
  store = store || useStore()
  const router = useRouter()
  const language = useGettext()
  const { $gettext, $ngettext, interpolate: $gettextInterpolate } = language
  const hasShareJail = useCapabilityShareJailEnabled()
  const hasSpacesEnabled = useCapabilitySpacesEnabled()
  const clientService = useClientService()
  const loadingService = useLoadingService()
  const { owncloudSdk } = clientService

  const queue = new PQueue({ concurrency: 4 })
  const deleteOps = []
  const resourcesToDelete = ref([])

  const isInTrashbin = computed(() => {
    return isLocationTrashActive(router, 'files-trash-generic')
  })

  const resources = computed(() => {
    return cloneStateObject(unref(resourcesToDelete))
  })

  const dialogTitle = computed(() => {
    const currentResources = unref(resources)
    const isFolder = currentResources[0].type === 'folder'
    let title = null

    if (currentResources.length === 1) {
      if (isFolder) {
        title = unref(isInTrashbin)
          ? $gettext('Permanently delete folder %{name}')
          : $gettext('Delete folder %{name}')
      } else {
        title = unref(isInTrashbin)
          ? $gettext('Permanently delete file %{name}')
          : $gettext('Delete file %{name}')
      }
      return $gettextInterpolate(
        title,
        {
          name: currentResources[0].name
        },
        true
      )
    }

    title = unref(isInTrashbin)
      ? $ngettext(
          'Permanently delete selected resource?',
          'Permanently delete %{amount} selected resources?',
          currentResources.length
        )
      : $ngettext(
          'Delete selected resource?',
          'Delete %{amount} selected resources?',
          currentResources.length
        )

    return $gettextInterpolate(title, { amount: currentResources.length }, false)
  })

  const dialogMessage = computed(() => {
    const currentResources = unref(resources)
    const isFolder = currentResources[0].type === 'folder'

    if (currentResources.length === 1) {
      if (isFolder) {
        return unref(isInTrashbin)
          ? $gettext(
              'Are you sure you want to delete this folder? All its content will be permanently removed. This action cannot be undone.'
            )
          : $gettext('Are you sure you want to delete this folder?')
      }
      return unref(isInTrashbin)
        ? $gettext(
            'Are you sure you want to delete this file? All its content will be permanently removed. This action cannot be undone.'
          )
        : $gettext('Are you sure you want to delete this file?')
    }

    return unref(isInTrashbin)
      ? $gettext(
          'Are you sure you want to delete all selected resources? All their content will be permanently removed. This action cannot be undone.'
        )
      : $gettext('Are you sure you want to delete all selected resources?')
  })

  const trashbin_deleteOp = (space, resource) => {
    const path = unref(hasShareJail)
      ? buildWebDavSpacesTrashPath(space.id)
      : buildWebDavFilesTrashPath(store.getters.user.id)

    return owncloudSdk.fileTrash
      .clearTrashBin(path, resource.id)
      .then(() => {
        store.dispatch('Files/removeFilesFromTrashbin', [resource])
        const translated = $gettext('"%{file}" was deleted successfully')
        store.dispatch('showMessage', {
          title: $gettextInterpolate(translated, { file: resource.name }, true)
        })
      })
      .catch((error) => {
        if (error.statusCode === 423) {
          // TODO: we need a may retry option ....
          const p = queue.add(() => {
            return trashbin_deleteOp(space, resource)
          })
          deleteOps.push(p)
          return
        }

        console.error(error)
        const translated = $gettext('Failed to delete "%{file}"')
        store.dispatch('showMessage', {
          title: $gettextInterpolate(translated, { file: resource.name }, true),
          status: 'danger'
        })
      })
  }

  const trashbin_delete = (space: SpaceResource) => {
    // TODO: use clear all if all files are selected
    store.dispatch('toggleModalConfirmButton')
    // FIXME: Implement proper batch delete and add loading indicator
    for (const file of unref(resources)) {
      const p = queue.add(() => {
        return trashbin_deleteOp(space, file)
      })
      deleteOps.push(p)
    }

    Promise.all(deleteOps).then(() => {
      store.dispatch('hideModal')
      store.dispatch('toggleModalConfirmButton')
    })
  }

  const filesList_delete = (space: SpaceResource) => {
    return loadingService.addTask(
      (loadingCallbackArgs) => {
        return store
          .dispatch('Files/deleteFiles', {
            ...language,
            space,
            files: unref(resources),
            clientService,
            loadingCallbackArgs
          })
          .then(async () => {
            store.dispatch('hideModal')
            store.dispatch('toggleModalConfirmButton')

            // Load quota
            if (
              isLocationSpacesActive(router, 'files-spaces-generic') &&
              !['public', 'share'].includes(space?.driveType)
            ) {
              if (unref(hasSpacesEnabled)) {
                const graphClient = clientService.graphAuthenticated
                const driveResponse = await graphClient.drives.getDrive(
                  unref(resources)[0].storageId
                )
                store.commit('runtime/spaces/UPDATE_SPACE_FIELD', {
                  id: driveResponse.data.id,
                  field: 'spaceQuota',
                  value: driveResponse.data.quota
                })
              } else {
                const user = await owncloudSdk.users.getUser(store.getters.user.id)
                store.commit('SET_QUOTA', user.quota)
              }
            }

            if (
              unref(resourcesToDelete).length &&
              isSameResource(unref(resourcesToDelete)[0], store.getters['Files/currentFolder'])
            ) {
              return router.push(
                createFileRouteOptions(space, {
                  path: dirname(unref(resourcesToDelete)[0].path),
                  fileId: unref(resourcesToDelete)[0].parentFolderId
                })
              )
            }
          })
      },
      { indeterminate: false }
    )
  }

  const deleteHelper = (space: SpaceResource) => {
    store.dispatch('toggleModalConfirmButton')

    unref(isInTrashbin) ? trashbin_delete(space) : filesList_delete(space)
  }

  const displayDialog = (space: SpaceResource, resources: Resource[]) => {
    resourcesToDelete.value = [...resources]

    const modal = {
      variation: 'danger',
      title: unref(dialogTitle),
      message: unref(dialogMessage),
      cancelText: $gettext('Cancel'),
      confirmText: $gettext('Delete'),
      onCancel: () => {
        store.dispatch('hideModal')
      },
      onConfirm: () => {
        deleteHelper(space)
      }
    }

    store.dispatch('createModal', modal)
  }

  return {
    displayDialog,
    // HACK: exported for unit tests:
    filesList_delete
  }
}
