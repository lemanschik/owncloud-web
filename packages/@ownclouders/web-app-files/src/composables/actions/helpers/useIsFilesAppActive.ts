import { computed, unref } from 'vue'
import { activeApp, useRoute } from 'web-pkg/src/composables'

const isFilesAppActive = (activeApp: string): boolean => {
  // FIXME: we should use this constant but it somehow breaks the unit tests
  // return activeApp === FilesApp.appInfo.id
  return activeApp === 'files'
}

export const useIsFilesAppActive = () => {
  const currentRoute = useRoute()

  return computed(() => isFilesAppActive(activeApp(unref(currentRoute))))
}
