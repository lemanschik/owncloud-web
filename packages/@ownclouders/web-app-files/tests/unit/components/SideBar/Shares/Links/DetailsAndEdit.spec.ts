import DetailsAndEdit from 'web-app-files/src/components/SideBar/Shares/Links/DetailsAndEdit.vue'
import { LinkShareRoles } from 'web-client/src/helpers/share'
import {
  createStore,
  defaultPlugins,
  shallowMount,
  defaultStoreMockOptions
} from 'web-test-helpers'
import { mockDeep } from 'jest-mock-extended'
import { Resource } from 'web-client'

const availableRoleOptions = LinkShareRoles.list(false, true, true, true)

const exampleLink = {
  name: 'Example link',
  url: 'https://some-url.com/abc',
  permissions: 1
}

describe('DetailsAndEdit component', () => {
  describe('if user can not edit', () => {
    it('does not render dropdown or edit button', () => {
      const { wrapper } = getShallowMountedWrapper(exampleLink)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('if user can edit', () => {
    it('renders dropdown and edit button', () => {
      const { wrapper } = getShallowMountedWrapper(exampleLink, false, true)
      expect(wrapper.html()).toMatchSnapshot()
    })

    it.todo('test edit options, button clicks and event handling/propagation')
  })

  describe('method "checkInputValue"', () => {
    it('should not show an error if value is valid', () => {
      const { wrapper } = getShallowMountedWrapper(exampleLink, false, true)
      const setModalInputErrorMessageStub = jest.spyOn(wrapper.vm, 'setModalInputErrorMessage')
      wrapper.vm.checkInputValue('New link name')
      expect(setModalInputErrorMessageStub).toHaveBeenCalledWith(null)
    })
    it('should show an error if value is longer than 255 characters', () => {
      const { wrapper } = getShallowMountedWrapper(exampleLink, false, true)
      const setModalInputErrorMessageStub = jest.spyOn(wrapper.vm, 'setModalInputErrorMessage')
      wrapper.vm.checkInputValue('n'.repeat(256))
      expect(setModalInputErrorMessageStub).toHaveBeenCalledWith(expect.anything())
    })
  })

  describe('method "checkPassword"', () => {
    it('should not show an error if value is valid', () => {
      const { wrapper } = getShallowMountedWrapper(exampleLink, false, true)
      const setModalInputErrorMessageStub = jest.spyOn(wrapper.vm, 'setModalInputErrorMessage')
      wrapper.vm.checkPassword('Password1234')
      expect(setModalInputErrorMessageStub).toHaveBeenCalledWith(null)
    })
    it('should show an error if value is longer than 72 characters', () => {
      const { wrapper } = getShallowMountedWrapper(exampleLink, false, true)
      const setModalInputErrorMessageStub = jest.spyOn(wrapper.vm, 'setModalInputErrorMessage')
      wrapper.vm.checkPassword('n'.repeat(73))
      expect(setModalInputErrorMessageStub).toHaveBeenCalledWith(expect.anything())
    })
  })
})

function getShallowMountedWrapper(link, expireDateEnforced = false, isModifiable = false) {
  const storeOptions = defaultStoreMockOptions
  const store = createStore(storeOptions)
  return {
    wrapper: shallowMount(DetailsAndEdit, {
      props: {
        availableRoleOptions,
        canRename: true,
        expirationDate: {
          enforced: expireDateEnforced,
          default: null,
          min: 'Wed Apr 01 2020 00:00:00 GMT+0000 (Coordinated Universal Time)',
          max: null
        },
        link,
        isModifiable,
        isPasswordEnforced: false,
        file: mockDeep<Resource>()
      },
      global: {
        renderStubDefaultSlot: true,
        stubs: { OcDatepicker: false, 'date-picker': true },
        plugins: [...defaultPlugins(), store]
      }
    })
  }
}
