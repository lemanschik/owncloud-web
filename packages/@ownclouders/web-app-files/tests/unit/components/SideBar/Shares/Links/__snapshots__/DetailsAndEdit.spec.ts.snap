// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsAndEdit component if user can edit renders dropdown and edit button 1`] = `
<div class="link-details oc-flex oc-flex-between oc-flex-middle" file="undefined">
  <div>
    <oc-button-stub appearance="raw" class="edit-public-link-role-dropdown-toggle oc-text-left" disabled="false" gapsize="none" id="edit-public-link-role-dropdown-toggle-undefined" justifycontent="center" size="medium" submit="button" type="button" variation="passive">
      <span class="link-current-role">Anyone with the link can view</span>
      <oc-icon-stub accessiblelabel="" color="" filltype="fill" name="arrow-down-s" size="medium" type="span" variation="passive"></oc-icon-stub>
    </oc-button-stub>
    <oc-drop-stub class="edit-public-link-role-dropdown" closeonclick="false" dropid="edit-public-link-role-dropdown" isnested="false" mode="click" offset="0" paddingsize="small" position="bottom-start" toggle="#edit-public-link-role-dropdown-toggle-undefined">
      <oc-list-stub class="role-dropdown-list" raw="false">
        <li>
          <oc-button-stub appearance="raw" class="oc-p-s" disabled="false" gapsize="medium" id="files-role-internal" justifycontent="space-between" size="medium" submit="button" type="button" variation="passive">
            <span class="oc-flex oc-flex-middle">
              <oc-icon-stub accessiblelabel="" class="oc-pl-s oc-pr-m" color="" filltype="fill" name="user" size="medium" type="span" variation="inherit"></oc-icon-stub>
              <span>
                <span class="oc-text-bold oc-display-block oc-width-1-1">Invited People</span>
                <span class="oc-text-small">Link works only for invited people. Login is required.</span>
              </span>
            </span>
            <span class="oc-flex">
              <!--v-if-->
            </span>
          </oc-button-stub>
        </li>
        <li>
          <oc-button-stub appearance="raw-inverse" class="selected oc-background-primary-gradient oc-p-s" disabled="false" gapsize="medium" id="files-role-viewer" justifycontent="space-between" size="medium" submit="button" type="button" variation="primary">
            <span class="oc-flex oc-flex-middle">
              <oc-icon-stub accessiblelabel="" class="oc-pl-s oc-pr-m" color="" filltype="fill" name="eye" size="medium" type="span" variation="inherit"></oc-icon-stub>
              <span>
                <span class="oc-text-bold oc-display-block oc-width-1-1">Can view</span>
                <span class="oc-text-small">Anyone with the link can view and download.</span>
              </span>
            </span>
            <span class="oc-flex">
              <oc-icon-stub accessiblelabel="" color="" filltype="fill" name="check" size="medium" type="span" variation="inherit"></oc-icon-stub>
            </span>
          </oc-button-stub>
        </li>
        <li>
          <oc-button-stub appearance="raw" class="oc-p-s" disabled="false" gapsize="medium" id="files-role-editor" justifycontent="space-between" size="medium" submit="button" type="button" variation="passive">
            <span class="oc-flex oc-flex-middle">
              <oc-icon-stub accessiblelabel="" class="oc-pl-s oc-pr-m" color="" filltype="fill" name="pencil" size="medium" type="span" variation="inherit"></oc-icon-stub>
              <span>
                <span class="oc-text-bold oc-display-block oc-width-1-1">Can edit</span>
                <span class="oc-text-small">Anyone with the link can view, download and edit.</span>
              </span>
            </span>
            <span class="oc-flex">
              <!--v-if-->
            </span>
          </oc-button-stub>
        </li>
      </oc-list-stub>
    </oc-drop-stub>
  </div>
  <div class="details-buttons">
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
    <div>
      <oc-button-stub appearance="raw" class="edit-drop-trigger" disabled="false" gapsize="medium" id="edit-public-link-dropdown-toggl-undefined" justifycontent="center" size="medium" submit="button" type="button" variation="passive">
        <oc-icon-stub accessiblelabel="" color="" filltype="fill" name="more-2" size="medium" type="span" variation="passive"></oc-icon-stub>
      </oc-button-stub>
      <oc-drop-stub closeonclick="false" dropid="edit-public-link-dropdown" isnested="false" mode="click" paddingsize="small" position="bottom-start" toggle="#edit-public-link-dropdown-toggl-undefined">
        <oc-list-stub class="edit-public-link-dropdown-menu oc-files-context-actions-border oc-pb-s" raw="false">
          <li class="oc-rounded oc-menu-item-hover">
            <oc-button-stub appearance="raw" class="oc-p-s action-menu-item" disabled="false" gapsize="medium" justifycontent="center" size="medium" submit="button" type="button" variation="passive">
              <oc-icon-stub accessiblelabel="" color="" filltype="line" name="pencil" size="medium" type="span" variation="passive"></oc-icon-stub>
              <span>Rename</span>
            </oc-button-stub>
          </li>
          <li class="oc-rounded oc-menu-item-hover">
            <date-picker-stub class="oc-datepicker link-expiry-picker link-expiry-picker" inputdebounce="1000" is24hr="false" isrange="false" locale="en" min-date="Wed Apr 01 2020 00:00:00 GMT+0000 (Coordinated Universal Time)" mode="date" modelconfig="[object Object]" popover="[object Object]" updateoninput="true">
              <!-- @slot Default slot to use as the popover anchor for datepicker -->
              <!-- args is undefined during initial render, hence we check it here -->
              <!--v-if-->
            </date-picker-stub>
          </li>
          <li class="oc-rounded oc-menu-item-hover">
            <oc-button-stub appearance="raw" class="oc-p-s action-menu-item" disabled="false" gapsize="medium" justifycontent="center" size="medium" submit="button" type="button" variation="passive">
              <oc-icon-stub accessiblelabel="" color="" filltype="line" name="lock-password" size="medium" type="span" variation="passive"></oc-icon-stub>
              <span>Add password</span>
            </oc-button-stub>
          </li>
        </oc-list-stub>
        <oc-list-stub class="edit-public-link-dropdown-menu oc-pt-s" raw="false">
          <li class="oc-rounded oc-menu-item-hover">
            <oc-button-stub appearance="raw" class="oc-p-s action-menu-item" disabled="false" gapsize="medium" justifycontent="center" size="medium" submit="button" type="button" variation="passive">
              <oc-icon-stub accessiblelabel="" color="" filltype="line" name="delete-bin-5" size="medium" type="span" variation="passive"></oc-icon-stub>
              <span>Delete link</span>
            </oc-button-stub>
          </li>
        </oc-list-stub>
      </oc-drop-stub>
    </div>
  </div>
</div>
`;

exports[`DetailsAndEdit component if user can not edit does not render dropdown or edit button 1`] = `
<div class="link-details oc-flex oc-flex-between oc-flex-middle" file="undefined">
  <p class="oc-my-rm">
    <span class="link-current-role">Anyone with the link can view</span>
  </p>
  <div class="oc-pr-s details-buttons">
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
  </div>
</div>
`;
